### DISCUSSION TITLE
Is there any way to embed JSX like DSL or equivalent in Gleam?

### DISCUSSION THREAD

**Discussion by mistyharsh** [gleam-lang/gleam#1448] [Questions & support] [answered]
Since Gleam now compiles to JavaScript, I would like to write bindings for Preact/React and use some syntactic sugar not necessarily same but like JSX/TSX to write HTML templates. Is there any way to do this in Gleam?

Also, can you please share what things do not work well and/or not supported when compiling the code to JavaScript?

**Comment by lpil** [comment, marked as answer]
Hello!

There's no XML literal syntax in Gleam, you'd have to use a function based API. One such example can be found here: https://github.com/JohnDoneth/gleam-vdom

> Also, can you please share what things do not work well and/or not supported when compiling the code to JavaScript?

What do you mean?

**Reply by mistyharsh** [reply]
Hey, thanks for the answer. About the other question, let me put it differently. How do I write following JS Code in Gleam? Sorry for newbie question, I could not find documentation:


```js

// 1. Tagged template literals
const css = tagged`
  width: 20px;
`;


// 2. Classes
class MyElm extends HTMLElement {

  constructor() {
    super();
    console.log('hello');
  }

}

// 3. Async code.
async function main() {
  const result = await someAsyncFunc();

  console.log(result);
}


// 4. Throw
function main() {
  throw 'hello';
}
```

**Reply by lpil** [reply]
You wouldn't, Gleam isn't an object oriented programming language, and it doesn't have throw, async/await, or tagged literals. As Gleam is a functional language you'd need to design a functional interface to any underlying JavaScript code and then use that in the Gleam code.

I've you've got some specific JavaScript APIs you might want to use from Gleam I could give some example of how they might be used.

**Reply by mistyharsh** [reply]
Writing a functional wrapper for classes to create web components is easy and I won't need `throw`. But, I am not sure how I can model async/await behavior. **Can you provide some samples for that?** For CSS-in-JS, I guess I can use object syntax instead of tagged templates.

Even better, a [cheat sheet](https://rescript-lang.org/docs/manual/latest/overview) like the one provided by ReScript for JS interop would be helpful.

So far, the experience has been very good. For now, I am exploring Pony language as a type-safe alternative to OTP/Elixir. But, I have really liked the design decisions you have made for the language so far. Following things are very good:

- FFI interface
- Strong type inference
- Union and opaque types
- My favorite `todo`
- Named arguments

After using Elm, I believe all the fancy stuff like Type classes, inheritance won't be necessary. But, yes, some form of meta-programming or scripting would really help.

**Reply by lpil** [reply]
> But, I am not sure how I can model async/await behavior. Can you provide some samples for that? 

We use a monadic promise API when running on JavaScript. Let me publish that package so you can view the docs. 

> Even better, a cheat sheet like the one provided by ReScript for JS interop would be helpful.

Totally, there's an issue for that. https://github.com/gleam-lang/gleam/issues/945

For now maybe one of the other cheatsheets could help: https://gleam.run/documentation/

Thank you for the kind words! I am also very interested in metaprogramming. Currently the focus is on tooling so it'll have to come later.

**Reply by lpil** [reply]
For now here are some tests using promises https://github.com/gleam-lang/javascript/blob/main/test/gleam/javascript/promise_test.gleam

**Reply by mistyharsh** [reply]
Exactly what I needed. Thanks a ton 👍 

Let me know if you need any help. Not a language designer but can definitely help with documentation or website.

**Reply by lpil** [reply]
Thank you, very kind. If you write any Gleam please do share it on GitHub and send me a link. Even if it's not finished it's great to see what people are doing, and it will help us get syntax highlighting on GitHub too once we have enough Gleam repos.

**Reply by andresgutgon** [reply]
Hi @lpil I'm coming from React. So far gleam and Lustre mental model are fantastic in terms of DX. My only grip is the view layer. Jsx-like would help a lot to frontend folks like me in terms of adoption imo

Have you seen?
https://github.com/choltreppe/gleamx/tree/main

Do you think this could be push by the core team?

**Reply by lpil** [reply]
We do not support any forks of the language, no. Ideally that language would remove the word "gleam" from the name to prevent people thinking it is Gleam instead of a new Gleam-inspired programming language.

**Reply by andresgutgon** [reply]
I respect that, but naming is not what matters to me 😆. Also, that package is not a fork. Is an AST on top of Gleam. The project is immature, but the idea is there. 

My question was more about what you think about the idea of Gleam or Lustre supporting in the future JSX-like syntax. This is the final piece for me to be fully in love. 

Not a deal breaker at all with AI I don't feel the pain of writing UI with Lustre but is hard to see coming from JSX

**Reply by lpil** [reply]
It is a fork of Gleam itself. It is a new Gleam-inspired language that has incompatible syntax with Gleam, and a new set of tooling that is distinct from Gleam's. It is not a fork of the Gleam compiler, but that is not required to fork a language.

No one has made a proposal for a new language feature in Gleam that would enable such a syntax. If you have an idea this post goes into detail about how to refine the idea and create a proposal: https://lpil.uk/blog/how-to-add-metaprogramming-to-gleam/ It talks about metaprogramming proposals specifically, but its advice is applicable to all language features.
